pallets/mips/src/lib.rs

```diff
@@ -640,37 +679,70 @@ decl_module! {
                 Error::<T>::NotACommitteeMember
             );
 
-            let mip = Self::proposals(index)
-                .ok_or(Error::<T>::MismatchedProposalIndex)?;
+            ensure!(Self::proposals(index).is_some(), Error::<T>::MismatchedProposalIndex);
 
             Self::create_referendum(
                 index,
-                MipsPriority::High,
-                mip.proposal,
+                ReferendumState::Pending,
+                ReferendumType::FastTracked,
             );
 
-            Self::deposit_event(RawEvent::ProposalFastTracked(index));
-            Self::close_proposal(index);
+            Self::update_proposal_state(index, ProposalState::Referendum);
+
+            // Update proposal metadata so we don't re-execute it later
+            <ProposalMetadata<T>>::mutate( |metas| {
+                let meta = metas.iter_mut().find( |meta| meta.index == index);
+                if let Some(meta) = meta {
+                    meta.end = Zero::zero();
+                }
+            });
+
+            Self::refund_proposal(index);
+            Ok(())
         }
 
         /// Governance committee can make a proposal that automatically becomes a referendum on
         /// which the committee can vote on.
         #[weight = SimpleDispatchInfo::FixedOperational(200_000)]
-        pub fn submit_referendum(origin, proposal: Box<T::Proposal>) {
-            let sender_key = AccountKey::try_from(ensure_signed(origin)?.encode())?;
-            let did = Context::current_identity_or::<Identity<T>>(&sender_key)?;
+        pub fn emergency_referendum(
+            origin,
+            proposal: Box<T::Proposal>,
+            url: Option<Url>,
+            description: Option<MipDescription>,
+        ) -> DispatchResult {
+            let proposer = ensure_signed(origin)?;
+            let proposer_key = AccountKey::try_from(proposer.encode())?;
+            let did = Context::current_identity_or::<Identity<T>>(&proposer_key)?;
 
             ensure!(
                 T::GovernanceCommittee::is_member(&did),
                 Error::<T>::NotACommitteeMember
             );
 
             let index = Self::next_index();
+            let mip = MIP {
+                index,
+                proposal: *proposal,
+                state: ProposalState::Referendum,
+            };
+            <Proposals<T>>::insert(index, mip);
+
+            let proposal_meta = MipsMetadata {
+                proposer: proposer.clone(),
+                index,
+                end: Zero::zero(),
+                url,
+                description,
+                cool_off_until: Zero::zero(),
+            };
+            <ProposalMetadata<T>>::mutate(|metadata| metadata.push(proposal_meta));
```

Comment:
Unbounded storage access. Ok for testnet v1, should revisit for v2. Perhaps add a TODO?

Reply:
Added TODO comment

Reply:
BTW, It has been fixed in [PR 303](https://github.com/PolymathNetwork/Polymesh/pull/303/)